File: Python3/601 - 800/673 Number of Longest Increasing Subsequence.py
# 584ms, 90.70%
class Solution:
    def findNumberOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dp = [[1, 1] for _ in range(len(nums))]
        max_for_all, res = 1, 0
        for i, num in enumerate(nums):
            max_len, count = 1, 0
            for j in range(i):
                if nums[j] < num:
                    if max_len == dp[j][0] + 1:
                        count += dp[j][1]
                    elif max_len < dp[j][0] + 1:
                        max_len = dp[j][0] + 1
                        count = dp[j][1]
            dp[i] = [max_len, max(count, dp[i][1])]
            max_for_all = max(max_len, max_for_all)
        return sum([item[1] for item in dp if item[0] == max_for_all])
    

File: 601 - 800/673 Number of Longest Increasing Subsequence.cc
// 34ms, 86.55%
static int x = []() {
    ios_base::sync_with_stdio(false); // toggles off the synchronization
    cin.tie(nullptr);                 // ties cin with nullptr, not cout
    return 0;
}();

class Solution
{
  public:
    int findNumberOfLIS(const vector<int> &nums)
    {
        if (nums.empty())
            return 0;
        int size = nums.size(), *dp[2];
        dp[0] = new int[size], dp[1] = new int[size];
        fill_n(dp[0], size, 1), fill_n(dp[1], size, 1);

        int max_for_all = 1, res = 0;
        for (int i = 0; i < size; ++i)
        {
            int max_len = 1, count = 0;
            for (int j = 0; j < i; ++j)
                if (nums[i] > nums[j])
                    if (max_len == dp[0][j] + 1)
                        count += dp[1][j];
                    else if (max_len < dp[0][j] + 1)
                    {
                        max_len = dp[0][j] + 1;
                        count = dp[1][j];
                    }
            dp[0][i] = max_len, dp[1][i] = count > dp[1][i] ? count : dp[1][i];
            max_for_all = max_len > max_for_all ? max_len : max_for_all;
        }

        for (int i = 0; i < size; ++i)
            if (dp[0][i] == max_for_all)
                res += dp[1][i];
        delete[] dp[0], delete[] dp[1];
        return res;
    }
};
